### PR TITLE
[Text Analytics] Mitigate transient test failures due to "InternalServerError" with 200 OK

### LINKED ISSUE #34970: [Text Analytics] Mitigate transient test failures due to "InternalServerError" with 200 OK
Occasionally during our test runs, the service responds back with a 200 OK response status but includes an "InternalServerError" in the response body for one or more text analysis tasks in that batch. In the case of LROs, the client cannot recover using its retry policy. Implement a solution to improve test reliability.

### PR DESCRIPTION
**Pull request by joseharriaga**
Resolves: https://github.com/Azure/azure-sdk-for-net/issues/34970

I'm adding a custom `NUnitAttribute` that I'm calling `RetryOnInternalServerErrorAttribute` and which is basically a simple modification of the [`RetryAttribute`](https://github.com/nunit/nunit/blob/30e1255feb2c7108e6b7b168126360ac3689d4bb/src/NUnitFramework/framework/Attributes/RetryAttribute.cs) in the NUnit framework. Its purpose is to specify that a test must be retried in the specific case of a known transient issue where the server returns a successful 200 OK response status but reports an internal server error in the response body for one or more text analysis tasks. If the test continues to fail due to this particular issue after reaching the max try count, the test result is marked as inconclusive.

### CODE REVIEW

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs:1]: Any reason not to add this to Azure.Core.TestFramework as I did with the IgnoreServerErrorAttribute? Could make the matching message customizable.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                        context.CurrentResult = innerCommand.Execute(context);
+                    }
+                    catch (Exception ex)
+                    {
+                        context.CurrentResult ??= context.CurrentTest.MakeTestResult();
+                        context.CurrentResult.RecordException(ex);
+                    }
+
+                    if (!ShouldRetry(context))
+                        break;
```
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs:74]: Stylistic: we prefer braces around everything. Was there not at least a warning on this? There used to be.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                    }
+                }
+
+                return context.CurrentResult;
+            }
+
+            /// <summary>
+            /// Indicates whether the encountered exception corresponds to the issue in question.
+            /// </summary>
+            private bool ShouldRetry(TestExecutionContext context)
```
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs:95]: Nit: could be static unless we did want to make it more general purpose.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/AbstractSummaryTests.cs [hunk 3]
```diff
@@ -76,6 +77,7 @@ public AbstractSummaryTests(bool isAsync, TextAnalyticsClientOptions.ServiceVers
         private const int AbstractiveSummarizationSentenceCount = 3;
 
         [RecordedTest]
+        [RetryOnInternalServerError]
```
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/AbstractSummaryTests.cs:80]: Can an `IRepeatTest` be on a class? Might be better to support either so it's usage isn't so verbose.

**Review by heaths — APPROVED** [review]

**Review by deyaaeldeen — APPROVED** [review]

**Review by mssfang — APPROVED** [review]

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/AbstractSummaryTests.cs (hunk 3 above)
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/AbstractSummaryTests.cs:80]: I wanted this too, but unfortunately, it only works with test methods. 😢

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/Infrastructure/RetryOnInternalServerErrorAttribute.cs:1]: I moved it to Azure.Core.TestFramework and made it customizable. Let me know what you think.

File: sdk/core/Azure.Core.TestFramework/src/RetryOnErrorAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        private readonly int _tryCount;
+
+        private readonly Func<TestExecutionContext, bool> _shouldRetry;
+
+        /// <summary>
+        /// Initializes a <see cref="RetryOnErrorAttribute"/> instance.
+        /// </summary>
+        /// <param name="tryCount">The number of times that the test can be executed.</param>
+        /// <param name="shouldRetry">The function that determines whether the test should be retried.</param>
+        public RetryOnErrorAttribute(int tryCount, Func<TestExecutionContext, bool> shouldRetry)
```
Comment (heaths) [inline sdk/core/Azure.Core.TestFramework/src/RetryOnErrorAttribute.cs:29]: This couldn't possibly be serialized as an attribute parameter, could it? If this is intended to be derived, I'd make the class abstract and this constructor protected. Or would the default implementation be good enough with just a `tryCount` to put in a public constructor, and still make this protected but keep the class instantiable?

**Review by heaths — APPROVED** [review]

**Pull request merged**